noir-r1cs/src/sparse_matrix.rs

```diff
@@ -175,10 +182,28 @@ impl Mul<HydratedSparseMatrix<'_>> for &[FieldElement] {
             rhs.matrix.num_rows,
             "Vector length does not match number of rows."
         );
+
+        let num_threads = rayon::current_num_threads();
         let mut result = vec![FieldElement::zero(); rhs.matrix.num_cols];
-        for ((i, j), value) in rhs.iter() {
-            result[j] += value * self[i];
-        }
+
+        let chunk_size = result.len().div_ceil(num_threads);
+
+        // In microbenchmarks par_iter_mut.chunks outperforms par_chunks_mut slightly.
+        result
+            .par_iter_mut()
+            .chunks(chunk_size)
+            .enumerate()
```

Comment:
If you do `.enumerate().par_chunks_mut(size)` you should already get the `base` value, which simplifies the code.

Reply:
Ah, that won't work because you can't chunk the resulting tuples. Nevermind!